### PR TITLE
Api provides client with a collection of all articles

### PR DESCRIPTION
**Pull request by christerforsgren91**
PT: https://www.pivotaltracker.com/story/show/176030919

### user Stories
```
As a an api
In order to present the client with all articles
I would like to send of a collection of articles from our database
```
### Changes proposed
-Add Articles model
-Add Article index
-Articles sorted by created_at
-Response with id, title, teaser, date
-Add serializer for article index

### What i have learned working on this feature
-Use serializer
-Sorting with loops
-Using strftime
-More knowledge about modals, controllers, seeds

### CODE REVIEW

File: app/controllers/api/articles_controller.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class Api::ArticlesController < ApplicationController
+  def index
+    articles = Article.all.sort_by(&:created_at).reverse
```
Comment (tochman) [inline app/controllers/api/articles_controller.rb]: This is a good example of logic that could be placed in the model and make use of scope. Not a blocker.

File: app/controllers/api/articles_controller.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+    if articles == []
+      render json: { articles: articles }, status: 204
```
Comment (tochman) [inline app/controllers/api/articles_controller.rb:5]: Okay, I'll buy that.

File: app/serializers/articles_index_serializer.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+class ArticlesIndexSerializer < ActiveModel::Serializer
+  attributes :id, :title, :teaser, :date
+
+  def date
+    object.created_at.strftime('%F')
```
Comment (tochman) [inline app/serializers/articles_index_serializer.rb:5]: Good. There is a potential problem with this solution, but that will manifest itself later if ever.

File: config/routes.rb [hunk 4]
```diff
@@ -1,3 +1,8 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  namespace :api do
+    get 'articles/index'
+  end
```
Comment (tochman) [inline config/routes.rb]: This is an auto-generated block of code. As we say in Swedish: "kaka på kaka" (see code below)

File: db/seeds.rb [hunk 5]
```diff
@@ -1,7 +1,16 @@
...
+"Amateur Rocket-Maker Finally Launches Himself Off Earth - Now To Prove It's Flat"
+]
+teasers = [
+  'Facebook has been criticized for failing to curb misinformation in English. But little attention has been paid to the scale of the problem in Arabic.',
+'Why fears of violence on March 4, the mythical day Trump was supposed to be inaugurated to a second term, proved unfounded.',
+'Happy New Year 2020: This list of ten moments includes not just personalities in India and across the world, but also certain events which have had a number of consequences for those involved.',
+'Mike Hughes, a California man who is most known for his belief that the Earth is shaped like a Frisbee, finally blasted off into the sky in a steam-powered rocket he had built himself.'
+]
+
+for i in 0...titles.count
```
Comment (tochman) [inline db/seeds.rb:14]: Wow! Why not keep it simple and use 

```ruby
titles.each do |title|
  # whatever you need to  do with each 'title'
end

File: spec/requests/api/can_responde_with_list_of_articles.spec.rb [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    end
+
+    it 'is expected to contain dates' do
+      expect(response_json['articles'].first['date']).to eq Time.now.strftime('%F')
+    end
+
+    it 'is expected to return lists with newest articles first' do
+      expect(response_json['articles'].first['title']).to eq 'Third Article'
+    end
+  end
```
Comment (tochman) [inline spec/requests/api/can_responde_with_list_of_articles.spec.rb:33]: I don't really see if you actually test the serialized output.... I can let this one pass, but I'm not sure if this way is sustainable in the long run.

File: spec/requests/api/can_responde_with_list_of_articles.spec.rb [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    it 'is expected to return 200 response status' do
+      expect(response).to have_http_status 200
+    end
+
+    it 'is expected to have lenght of 3 articles' do
+      expect(response_json['articles'].count).to eq 3
+    end
+
+    it 'is expected to have a title' do
+      expect(response_json['articles'].first['title']).to eq 'Third Article'
```
Comment (tochman) [inline spec/requests/api/can_responde_with_list_of_articles.spec.rb:19]: `.first` is supposed to be 'Third Article'. Seems weird for an untrained eye.... I do understand that you are testing the sorting by date, but it does not come across....

**Review by tochman — CHANGES_REQUESTED** [review]
Nothing heavy, but not quite ready for prime-time.

**Comment by tochman** [comment]
Also, I would like you to make use of "What I learned" in the PR description. Your learning is not yet over.

File: db/seeds.rb (hunk 5 above)
Comment (LucasKnudsen) [reply db/seeds.rb:14]: We would need the index to access teaser, but could use `titles.each_with_index` for that, indeed!

File: spec/requests/api/can_responde_with_list_of_articles.spec.rb (hunk 6 above)
Comment (christerforsgren91) [reply spec/requests/api/can_responde_with_list_of_articles.spec.rb:33]: We tested date, but its true! we could have tested that updated at is not present.

File: db/seeds.rb (hunk 5 above)
Comment (LucasKnudsen) [reply db/seeds.rb:14]: Also we suck at `for` loops, so good practice 😄

File: config/routes.rb (hunk 4 above)
Comment (christerforsgren91) [reply config/routes.rb]: Removed

File: app/controllers/api/articles_controller.rb (hunk 1 above)
Comment (christerforsgren91) [reply app/controllers/api/articles_controller.rb]: fixed

File: spec/requests/api/can_responde_with_list_of_articles.spec.rb (hunk 7 above)
Comment (christerforsgren91) [reply spec/requests/api/can_responde_with_list_of_articles.spec.rb:19]: we edit the description.

File: db/seeds.rb (hunk 5 above)
Comment (tochman) [reply db/seeds.rb:14]: nah... you would have teaser in the `title['teaser']` won't you? 

sucking at for loops is not a bad thing... if you can avoid that syntax. `.each` IS a for loop

Comment (tochman) [reply db/seeds.rb:14]: Forget it. I see what you are doing now.... Okay. But mark my words, this is NOT a low-hanging fruit.

**Review by tochman — APPROVED** [review]
Okay boyz, this one can fly now.

**Pull request merged**